services/api/src/game/__tests__/game.service.test.ts

```diff
@@ -1,48 +1,166 @@
-  row: 1,
-  tile: 1,
-  figure: 'rock',
-  player: 'test',
-  id: 'testId',
-  score: 1,
-};
+jest.mock('ws');
+jest.mock('../../logger');
 
 describe('game.service', () => {
   const roomId = 'abcd-1234';
-  const turns: PlaceFigureMessage[] = [];
+  const playerTestId = 'some-short-v4-uuid-1';
+  const testTurn: PlaceFigureMessage = {
+    row: 1,
+    tile: 1,
+    figure: 'rock',
+    player: 'player1',
+    id: playerTestId,
+    score: 1,
+  };
+  const anotherTestTurn: PlaceFigureMessage = {
+    row: 4,
+    tile: 4,
+    figure: 'rock',
+    player: 'player1',
+    id: playerTestId,
+    score: 1,
+  };
+
+  beforeAll(() => {
+    gameRoomService.getOrCreateRoom(roomId);
+    jest.spyOn(global.Math, 'random').mockReturnValue(1);
+    Object.defineProperty(ws, 'readyState', { value: WebSocket.OPEN });
+  });
+
+  afterEach(() => {
+    jest.clearAllMocks();
+    playerService.unsubscribe(ws);
+    gameRoomService.getTurns(roomId).length = 0;
+  });
+
+  it('should return false, when player has not moved, after the move should return true', () => {
+    const connectedVoter = voterConnect();
+    const doesPlayerHaveMove = gameService.playerHasMove(
+      roomId,
+      connectedVoter.id,
+    );
+    expect(doesPlayerHaveMove).toBeFalsy();
+    const message: ReceivedMessage<MessageType.FigureMoved> = {
+      type: MessageType.FigureMoved,
+      payload: testTurn,
+    };
+    playerService.newMessageReceived(ws, message);
+    const doesPlayerHaveMoveAfterMove = gameService.playerHasMove(
+      roomId,
+      connectedVoter.id,
+    );
+    expect(doesPlayerHaveMoveAfterMove).toBeTruthy();
+  });
 
-  beforeEach(() => {
-    (gameRoomService.getTurns as Mock).mockReturnValue(turns);
+  it('should return all turns', () => {
+    const figureMovedSpy = jest.spyOn(gameService, 'figureMoved');
+    voterConnect();
+    const turnsCount = gameRoomService.getTurns(roomId).length;
+    expect(turnsCount).toBe(0);
+    const message: ReceivedMessage<MessageType.FigureMoved> = {
+      type: MessageType.FigureMoved,
+      payload: testTurn,
+    };
+    playerService.newMessageReceived(ws, message);
+    const turnsAfterMove = gameRoomService.getTurns(roomId).length;
+    expect(turnsAfterMove).toBe(1);
+    expect(figureMovedSpy).toReturnWith([testTurn]);
   });
 
-  afterEach(jest.clearAllMocks);
+  it('should return a turn by player id', () => {
+    voterConnect();
+    const message: ReceivedMessage<MessageType.FigureMoved> = {
+      type: MessageType.FigureMoved,
+      payload: testTurn,
+    };
+    playerService.newMessageReceived(ws, message);
+    const playerTurn = gameService.findMoveByPlayerId(roomId, playerTestId);
+    expect(playerTurn).toMatchObject(testTurn);
+  });
+
+  it('should remove players turn', () => {
+    voterConnect();
+    const message: ReceivedMessage<MessageType.FigureMoved> = {
+      type: MessageType.FigureMoved,
+      payload: testTurn,
+    };
+    playerService.newMessageReceived(ws, message);
```

Comment:
Since this is used all over the place, can we extract it into a separate method?